src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetQueryService.cs

```diff
@@ -0,0 +1,515 @@
+
+        foreach (var locationLevel in locationLevels)
+        {
+            fieldColumnMap[LocationsSortField(locationLevel)] = DataTable.Ref().LocationId(locationLevel);
+        }
+
+        var otherFields = columns
+            .Except([
+                DataTable.Cols.Id,
+                ..ReservedSorts.Keys,
+                ..locationLevels.Select(DataTable.Cols.LocationId),
+            ])
+            .Concat(indicators)
+            .ToHashSet();
+
+        var invalidSorts = new HashSet<DataSetQuerySort>();
+        var sorts = new List<Sort>();
+
+        foreach (var sort in request.Sorts)
+        {
+            if (fieldColumnMap.TryGetValue(sort.Field, out var field))
+            {
+                sorts.Add(new Sort(Field: field, Direction: sort.ParsedDirection));
+                continue;
+            }
+
+            if (otherFields.Contains(sort.Field))
+            {
+                sorts.Add(new Sort(Field: sort.Field, Direction: sort.ParsedDirection));
+                continue;
+            }
+
+            invalidSorts.Add(sort);
+        }
+
+        if (invalidSorts.Count != 0)
+        {
+            queryState.Errors.Add(new ErrorViewModel
+            {
+                Message = ValidationMessages.SortsNotFound.Message,
+                Code = ValidationMessages.SortsNotFound.Code,
+                Path = "sorts",
+                Detail = new NotFoundItemsErrorDetail<DataSetQuerySort>(invalidSorts)
+            });
+        }
+
+        return sorts;
+    }
+
+    private static ColumnsByType GetColumnsByType(
+        ISet<string> columns,
+        ISet<string> allIndicators,
+        ISet<string> selectedIndicators)
+    {
+        var locationLevels = GeographicLevelUtils.Levels
+            .Where(level => columns.Contains(DataTable.Cols.LocationId(level)))
+            .ToHashSet();
+
+        var filterColumns = columns
+            .Except([
+                ..ReservedColumns,
+                ..locationLevels.Select(DataTable.Cols.LocationId),
+                ..allIndicators
+            ])
+            .ToHashSet();
+
+        return new ColumnsByType
+        {
+            Filters = filterColumns,
+            LocationLevels = locationLevels,
+            Indicators = selectedIndicators
+        };
+    }
+
+    private async Task<List<DataSetQueryResultViewModel>> MapQueryResults(
+        IEnumerable<IDictionary<string, object?>> rows,
+        DataSetVersion dataSetVersion,
+        ColumnsByType columnsByType,
+        bool debug,
+        CancellationToken cancellationToken)
+    {
+        using var _ = MiniProfiler.Current.Step(
+            $"{nameof(DataSetQueryService)}.{nameof(MapQueryResults)}");
+
+        var locationColumnsByCode = columnsByType.LocationLevels
+            .ToDictionary(
+                level => level.GetEnumValue(),
+                DataTable.Cols.LocationId
+            );
+
+        var rowList = new List<IDictionary<string, object?>>();
```

Comment:
Can we remove `rowList` and just use `rows` on line 362?

Reply:
This was to avoid multiple enumeration warnings in Rider as we're working with `IEnumerable` and I didn't want to have allocate 

However, I've now changed the repository methods to return `IList` instead of `IEnumerable` so we don't need this any more.